src/nucleotide.rs

```diff
@@ -2,83 +2,203 @@ use std::fmt::{self, Write};
-        insert!(b's', Self::N); // C or G (complement S (self))
-        insert!(b'y', Self::N); // C or T (complement R)
-        insert!(b'k', Self::N); // G or T (complement M)
-        insert!(b'v', Self::N); // A, C or G (complement B)
-        insert!(b'h', Self::N); // A, C or T (complement D)
-        insert!(b'd', Self::N); // A, G or T (complement H)
-        insert!(b'b', Self::N); // C, G or T (complement V)
+    fn to_ascii(self) -> u8 {
+        match self {
+            Self::A => b'A',
+            Self::T => b'T',
+            Self::C => b'C',
+            Self::G => b'G',
+        }
+    }
 
-        pack_table
+    fn is_ambiguous(self) -> bool {
+        false
     }
+}
 
-    const ASCII_PACK_TABLE: [u8; 128] = Self::ascii_pack_table();
+impl NucleotideAmbiguous {
+    pub const ALL: [Self; 15] = [
+        Self::A,
+        Self::T,
+        Self::W,
+        Self::C,
+        Self::M,
+        Self::Y,
+        Self::H,
+        Self::G,
+        Self::R,
+        Self::K,
+        Self::D,
+        Self::S,
+        Self::V,
+        Self::B,
+        Self::N,
+    ];
 
-    const COMPLEMENT_TABLE: [Self; Self::N_NUCLEOTIDES as usize] =
-        [Self::T, Self::A, Self::G, Self::C, Self::N];
-    pub const fn complement(self) -> Self {
-        Self::COMPLEMENT_TABLE[self as u8 as usize]
+    pub const fn possibilities(self) -> &'static [Nucleotide] {
+        match self {
+            Self::A => &[Nucleotide::A],
+            Self::T => &[Nucleotide::T],
+            Self::C => &[Nucleotide::C],
+            Self::G => &[Nucleotide::G],
+            Self::W => &[Nucleotide::A, Nucleotide::T],
+            Self::M => &[Nucleotide::A, Nucleotide::C],
+            Self::R => &[Nucleotide::A, Nucleotide::G],
+            Self::Y => &[Nucleotide::T, Nucleotide::C],
+            Self::S => &[Nucleotide::C, Nucleotide::G],
+            Self::K => &[Nucleotide::T, Nucleotide::G],
+            Self::B => &[Nucleotide::T, Nucleotide::C, Nucleotide::G],
+            Self::V => &[Nucleotide::A, Nucleotide::C, Nucleotide::G],
+            Self::D => &[Nucleotide::A, Nucleotide::T, Nucleotide::G],
+            Self::H => &[Nucleotide::A, Nucleotide::T, Nucleotide::C],
+            Self::N => &[Nucleotide::A, Nucleotide::T, Nucleotide::C, Nucleotide::G],
+        }
+    }
+}
+
+impl NucleotideLike for NucleotideAmbiguous {
+    fn complement(self) -> Self {
+        match self {
+            Self::A => Self::T,
+            Self::T => Self::A,
+            Self::W => Self::W,
+            Self::C => Self::G,
+            Self::M => Self::K,
+            Self::Y => Self::R,
+            Self::H => Self::D,
+            Self::G => Self::C,
+            Self::R => Self::Y,
+            Self::K => Self::M,
+            Self::D => Self::H,
+            Self::S => Self::S,
+            Self::V => Self::B,
+            Self::B => Self::V,
+            Self::N => Self::N,
+        }
+    }
```

Comment:
I ran some benchmarks, and it turns out that writing methods like this is faster than defining a static array and indexing into it with `self as usize`. (The generated code is almost the same, just without a bounds check. I guess casting to `usize` makes Rust doubt/forget that the indexing is sane.)

However for "sparse" conversions like ASCII-to-nucleotide, defining a lookup table with 256 entries is still faster, because it doesn't want to make one of those automatically. (I bumped the size of the table from 128 up to 256 because this eliminates a bounds check.)